test/cpp/api/serialize.cpp

```diff
@@ -600,6 +600,56 @@ TEST(SerializeTest, Optim_Adam) {
   is_optimizer_state_equal<AdamParamState>(optim1.state(), optim1_2.state());
 }
 
+TEST(SerializeTest, Optim_AdamW) {
+  // test_serialize_optimizer<AdamW, AdamWOptions, AdamWParamState>(AdamWOptions().lr(0.99999).amsgrad(true).weight_decay(0.5));
```

Comment:
This might be a precision issue, probably the testing parameter is not good enough to identify the difference. Can you try different parameters?

Reply:
Yes, and that is the case for e.g. (original) Adam with no weight decay, but it seems the right hyper-parameters are used in order for the test to pass (i.e. in `test/cpp/api/serialize.cpp:555` `weight_decay` is set to `0.5`). I ran a grid search and found the parameters that will make actually AdamW produce different outputs (hint: low `beta2`). It seems we have two options:

- Use the ad-hoc hyper-parameters that will produce different outpus
- Modify all tests to use the default hyper-parameters and modify the `test_serialize_optimizer` api to check for *stateful optimizers* or something simillar.

For now, I modified the test hyper-parameters in order to pass. If we want changes in `test_serialize_optimizer`, let's open another issue and I can work on it!

Reply:
I think for now, we can use ad-hoc hyper-parameters. Your change looks good.
The test_serialize_optimizer need some tweaks, it should check predictable state only. 
I will check how we want to refactor it and file issue later if necessary.

Reply:
Cool - if you figure it out and need a hand with that drop a mention!